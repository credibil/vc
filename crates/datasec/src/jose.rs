//! # JSON Object Signing and Encryption (JOSE) Proofs
//!
//! [JOSE] proofs are enveloping proofs for Credentials based on JWT [RFC7519],
//! JWS [RFC7515], and JWK [RFC7517].
//!
//! The Securing Verifiable Credentials using JOSE and COSE [VC-JOSE-COSE]
//! recommendation defines a "bridge" between these and the Verifiable
//! Credentials Data Model v2.0, specifying the suitable header claims, media
//! types, etc.
//!
//! In the case of JOSE, the Credential is the "payload". This is preceded by a
//! suitable header whose details are specified by Securing Verifiable
//! Credentials using JOSE and COSE for the usage of JWT. These are encoded,
//! concatenated, and signed, to be transferred in a compact form by one entity
//! to an other (e.g., sent by the holder to the verifier). All the intricate
//! details on signatures, encryption keys, etc., are defined by the IETF
//! specifications; see Example 6 for a specific case.
//!
//! ## Note
//!
//! If the JWT is only a JWE, iss, exp and aud MUST be omitted in the JWT Claims
//! Set of the JWE, and the processing rules as per JARM Section 2.4 related to
//! these claims do not apply. [OpenID4VP] JWT - JWE
//!
//! ```json
//! {
//!   "vp_token": "eyJI...",
//!   "presentation_submission": {...}
//! }
//! ```
//!
//! [JOSE]: https://datatracker.ietf.org/wg/jose/about
//! [RFC7515]: https://www.rfc-editor.org/rfc/rfc7515
//! [RFC7517]: https://www.rfc-editor.org/rfc/rfc7517
//! [RFC7519]: https://www.rfc-editor.org/rfc/rfc7519
//! [VC-JOSE-COSE]: https://w3c.github.io/vc-jose-cose
//! [OpenID4VP]: https://openid.net/specs/openid-4-verifiable-presentations-1_0.html

pub mod jwa;
pub mod jwe;
pub mod jwk;
pub mod jws;
pub mod jwt;
