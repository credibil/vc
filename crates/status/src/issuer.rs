//! # Status (Issuer)
//!
//! Traits and types for managing the publication of a credential status as an
//! issuer.

use std::future::Future;

pub use vercre_core::Quota;
pub use vercre_w3c_vc::model::CredentialStatus;

use crate::provider;

/// The `Status` trait specifies how status information can be looked up for a
/// credential.
pub trait Status: Send + Sync {
    /// Returns information on how to look up the status of a credential.
    ///
    /// A default implementation is provided that just returns `None` for cases
    /// where the issuer will not be providing a status endpoint.
    fn status(
        &self, _holder_subject: &str, _credential_identifier: &str,
    ) -> impl Future<Output = provider::Result<Option<Quota<CredentialStatus>>>> + Send {
        async { Ok(None) }
    }
}
