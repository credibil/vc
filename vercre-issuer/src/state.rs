//! State is used by the library to persist request information between steps
//! in the issuance process.

use std::collections::HashMap;

use chrono::{DateTime, TimeDelta, Utc};
use serde::{Deserialize, Serialize};
use vercre_openid::issuer::{
    AuthorizationDetail, CodeChallengeMethod, CredentialOffer, CredentialRequest, RequestObject,
};
use vercre_w3c_vc::model::VerifiableCredential;

type CredentialIdentifier = String;

/// State is used to persist request information between issuance steps in the
/// Credential issuance process.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct State {
    /// Identifies the (previously authenticated) Holder in order that Issuer
    /// can authorize credential issuance.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subject_id: Option<String>,

    /// Stage holds data relevant to the current state of the issuance process.
    /// This data is used by subsequent step(s) to verify Wallet interactions,
    /// including credential issuance.
    pub stage: Stage,

    /// Time state should expire.
    pub expires_at: DateTime<Utc>,
}

impl State {
    /// Determines whether state has expired or not.
    pub fn is_expired(&self) -> bool {
        self.expires_at.signed_duration_since(Utc::now()).num_seconds() < 0
    }
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
#[allow(clippy::large_enum_variant)]
pub enum Stage {
    /// Unauthorized state.
    #[default]
    Unauthorized,

    /// Holds a Credential Offer awaiting retrieval by the Wallet. The Wallet
    /// has been sent a unique URL it can use to retrieve the offer.
    Pending(CredentialOffer),

    /// Holds pre-authorized offer data as presented to the Wallet. This data is
    /// used when validating the Wallet's request for an access token.
    Offered(Offer),

    /// Holds a Pushed Authorization Request awaiting retrieval by the Wallet.
    PushedAuthorization(PushedAuthorization),

    /// Holds authorization data in cases where the Wallet requests and is
    /// granted authorization to request credential issuance. As with
    /// `PreAuthorized` state, this data is used when validating the
    /// Wallet's request for an access token.
    Authorized(Authorization),

    /// Holds information about the access token and corresponding credentials
    /// the Wallet is authorized to request.
    Validated(Token),

    /// Issued Credential state.
    Issued(Credential),

    /// Deferred issuance state.
    Deferred(Deferrance),
}

/// Pre-authorization state from the `create_offer` endpoint.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Offer {
    /// A list of `authorization_details` entries referencing credentials the
    /// Wallet is authorized to request.
    pub items: Option<Vec<AuthorizedItem>>,

    /// Transaction code sent to the holder to use (if present)when requesting
    /// an access token.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tx_code: Option<String>,
}

/// Holds data used during the issuance of a credential.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Authorized {
    /// Identifies the dataset associated with the credential to be issued.
    /// Dataset is unique by issuer not by subject.
    ///
    /// For example, the `credential_configuration_id` is `UniversityDegree_JWT`
    /// and the `credential_identifier` is `EngineeringDegree2023`.
    pub credential_identifier: String,

    /// Credential's `credential_configuration_id` connecting it with supported
    /// credential metadata.
    pub credential_configuration_id: String,

    /// Identifies a subset of claims to use when issuing the associated
    /// credential. This subset is used in cases where the Wallet has
    /// requested (and has been authorized for) issuance of a credential
    /// containing subset of claims.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub claim_ids: Option<Vec<String>>,
}

/// Authorization state.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
#[allow(clippy::struct_field_names)]
pub struct Authorization {
    /// The `client_id` of the Wallet requesting issuance.
    pub client_id: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub redirect_uri: Option<String>,

    /// PKCE code challenge from the Authorization Request.
    pub code_challenge: String,

    /// PKCE code challenge method from the Authorization Request.
    pub code_challenge_method: CodeChallengeMethod,

    /// A list of authorized `scope` or `authorization_details` entries along
    /// with credential metadata and dataset identifiers.
    pub items: Vec<AuthorizedItem>,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct PushedAuthorization {
    /// The Authorization Request pushed to the PAR endpoint.
    pub request: RequestObject,

    /// The time the request URI should expire at.
    pub expires_at: DateTime<Utc>,
}

/// Authorized `authorization_detail` or `scope` item along with
/// `credential_configuration_id` and `credential_identifier`s.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct AuthorizedItem {
    /// Authorized item.
    #[serde(flatten)]
    pub item: ItemType,

    /// Credential configuration metadata for the item.
    pub credential_configuration_id: String,

    /// Authorized credential datasets for the item.
    pub credential_identifiers: Vec<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
pub enum ItemType {
    /// Authorized item is of type `authorization_detail`
    AuthorizationDetail(AuthorizationDetail),

    /// Authorized item is of type `scope`
    Scope(String),
}

impl Default for ItemType {
    fn default() -> Self {
        Self::AuthorizationDetail(AuthorizationDetail::default())
    }
}

/// Token state.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Token {
    /// The access token.
    #[allow(clippy::struct_field_names)]
    pub access_token: String,

    /// The nonce to be used by the Wallet when creating a proof of possession
    /// of the key proof.
    pub c_nonce: String,

    /// Number denoting the lifetime in seconds of the `c_nonce`.
    pub c_nonce_expires_at: DateTime<Utc>,

    /// Credentials (configuration id and identifier) validated for issuance
    /// using the accompanying access token.
    pub credentials: HashMap<CredentialIdentifier, Authorized>,
}

impl Token {
    pub fn c_nonce_expires_in(&self) -> i64 {
        self.c_nonce_expires_at.signed_duration_since(Utc::now()).num_seconds()
    }

    pub fn c_nonce_expired(&self) -> bool {
        self.c_nonce_expires_in() < 0
    }
}

/// Issued Credential state (for Notification endpoint).
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Credential {
    pub credential: VerifiableCredential,
}

/// Deferred issuance state.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Deferrance {
    /// Used to identify a Deferred Issuance transaction. Is used as the
    /// state persistence key.
    pub transaction_id: String,

    /// Save the Credential request when issuance is deferred.
    pub credential_request: CredentialRequest,
}

pub enum Expire {
    Authorized,
    Access,
    Nonce,
}

impl Expire {
    pub fn duration(&self) -> TimeDelta {
        match self {
            Self::Authorized => TimeDelta::try_minutes(5).unwrap_or_default(),
            Self::Access => TimeDelta::try_minutes(15).unwrap_or_default(),
            Self::Nonce => TimeDelta::try_minutes(10).unwrap_or_default(),
        }
    }
}
