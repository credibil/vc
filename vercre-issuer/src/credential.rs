//! # Batch Credential Endpoint
//!
//! The Batch Credential Endpoint issues multiple Credentials in one Batch Credential
//! Response as approved by the End-User upon presentation of a valid Access Token
//! representing this approval.
//!
//! A Wallet can request issuance of multiple Credentials of certain types and formats
//! in one Batch Credential Request. This includes Credentials of the same type and
//! multiple formats, different types and one format, or both.

use std::fmt::Debug;

use chrono::Utc;
use tracing::instrument;
use vercre_core::{gen, Kind, Quota};
use vercre_datasec::jose::jws::{self, KeyType, Type};
use vercre_datasec::SecOps;
use vercre_openid::issuer::{
    CredentialConfiguration, CredentialDefinition, CredentialRequest, CredentialResponse,
    CredentialType, Issuer, Metadata, ProofClaims, ProofOption, ProofType, ProofsType, Provider,
    StateStore, Subject,
};
use vercre_openid::{Error, Result};
use vercre_w3c_vc::model::{CredentialSubject, VerifiableCredential};
use vercre_w3c_vc::proof::{Format, Payload};
use vercre_w3c_vc::verify_key;

use crate::state::{Deferred, Expire, State, Step};

/// Credential request handler.
///
/// # Errors
///
/// Returns an `OpenID4VP` error if the request is invalid or if the provider is
/// not available.
#[instrument(level = "debug", skip(provider))]
pub async fn credential(
    provider: impl Provider, request: &CredentialRequest,
) -> Result<CredentialResponse> {
    let Ok(buf) = StateStore::get(&provider, &request.access_token).await else {
        return Err(Error::AccessDenied("invalid access token".into()));
    };
    let Ok(state) = State::try_from(buf) else {
        return Err(Error::AccessDenied("invalid state for access token".into()));
    };

    let mut ctx = Context {
        state,
        issuer_config: Metadata::issuer(&provider, &request.credential_issuer)
            .await
            .map_err(|e| Error::ServerError(format!("metadata issue: {e}")))?,
        holder_did: String::new(),
    };

    verify(&mut ctx, provider.clone(), request).await?;
    process(&ctx, provider, request).await
}

#[derive(Debug)]
struct Context {
    issuer_config: Issuer,
    state: State,
    holder_did: String,
}

#[allow(clippy::too_many_lines)]
async fn verify(
    context: &mut Context, provider: impl Provider, request: &CredentialRequest,
) -> Result<()> {
    tracing::debug!("credential::verify");

    let Step::Token(token_state) = &context.state.current_step else {
        return Err(Error::AccessDenied("invalid access token state".into()));
    };

    // c_nonce expiry
    if token_state.c_nonce_expired() {
        return Err(Error::AccessDenied("c_nonce has expired".into()));
    }

    let mut supported_proofs = None;

    // TODO: add support for `credential_identifier`

    // format and type request
    if let CredentialType::Format(format) = &request.credential_type {
        let Some(definition) = &request.credential_definition else {
            return Err(Error::InvalidCredentialRequest("credential definition not set".into()));
        };

        // check request has been authorized:
        //   - match format + type against authorized items in state
        let mut authorized = false;

        for (identifier, config) in &context.issuer_config.credential_configurations_supported {
            if (&config.format == format)
                && (config.credential_definition.type_ == definition.type_)
            {
                supported_proofs = config.proof_types_supported.as_ref();
                authorized = context.state.credential_identifiers.contains(identifier);
                break;
            }
        }

        if !authorized {
            return Err(Error::InvalidCredentialRequest(
                "Requested credential has not been authorized".into(),
            ));
        }
    };

    // TODO: refactor into separate function.
    if let Some(supported_types) = supported_proofs {
        let Some(proof_option) = &request.proof_option else {
            return Err(Error::InvalidCredentialRequest("proof not set".into()));
        };

        // TODO: recheck this list for compliance
        // To validate a key proof, ensure that:
        //   - all required claims for that proof type are contained as defined
        //     in Section 7.2.1
        //   - the key proof is explicitly typed using header parameters as
        //     defined for that proof type
        //   - the header parameter indicates a registered asymmetric digital
        //     signature algorithm, alg parameter value is not none, is supported
        //     by the application, and is acceptable per local policy
        //   - the signature on the key proof verifies with the public key
        //     contained in the header parameter
        //   - the header parameter does not contain a private key
        //   - the nonce claim (or Claim Key 10) matches the server-provided
        //     c_nonce value, if the server had previously provided a c_nonce
        //   - the creation time of the JWT, as determined by either the issuance
        //      time, or a server managed timestamp via the nonce claim, is within
        //      an acceptable window (see Section 11.5).

        // TODO: cater for non-JWT proofs
        let _ = supported_types
            .get("jwt")
            .ok_or_else(|| Error::InvalidCredentialRequest("proof type not supported".into()))?;

        // extract proof JWT(s) from request
        let proof_jwts = match proof_option {
            ProofOption::Proof { proof_type } => match proof_type {
                ProofType::Jwt { jwt } => &vec![jwt.clone()],
            },
            ProofOption::Proofs(proofs_type) => match proofs_type {
                ProofsType::Jwt(proof_jwts) => proof_jwts,
            },
        };

        for proof_jwt in proof_jwts {
            // TODO: check proof is signed with supported algorithm (from proof_type)
            let jwt: jws::Jwt<ProofClaims> =
                match jws::decode(proof_jwt, verify_key!(&provider)).await {
                    Ok(jwt) => jwt,
                    Err(e) => {
                        let (c_nonce, c_nonce_expires_in) = err_nonce(context, &provider).await?;
                        return Err(Error::InvalidProof {
                            hint: format!("issue decoding JWT: {e}"),
                            c_nonce,
                            c_nonce_expires_in,
                        });
                    }
                };
            // proof type
            if jwt.header.typ != Type::Proof {
                let (c_nonce, c_nonce_expires_in) = err_nonce(context, &provider).await?;
                return Err(Error::InvalidProof {
                    hint: format!("Proof JWT 'typ' is not {}", Type::Proof),
                    c_nonce,
                    c_nonce_expires_in,
                });
            }

            // previously issued c_nonce
            if jwt.claims.nonce.as_ref() != Some(&token_state.c_nonce) {
                let (c_nonce, c_nonce_expires_in) = err_nonce(context, &provider).await?;
                return Err(Error::InvalidProof {
                    hint: "Proof JWT nonce claim is invalid".into(),
                    c_nonce,
                    c_nonce_expires_in,
                });
            }

            // TODO: use `decode` method in w3c-vc
            // Key ID
            let KeyType::KeyId(kid) = &jwt.header.key else {
                let (c_nonce, c_nonce_expires_in) = err_nonce(context, &provider).await?;

                return Err(Error::InvalidProof {
                    hint: "Proof JWT 'kid' is missing".into(),
                    c_nonce,
                    c_nonce_expires_in,
                });
            };
            // HACK: save extracted DID for later use when issuing credential
            let Some(did) = kid.split('#').next() else {
                let (c_nonce, c_nonce_expires_in) = err_nonce(context, &provider).await?;

                return Err(Error::InvalidProof {
                    hint: "Proof JWT DID is invalid".into(),
                    c_nonce,
                    c_nonce_expires_in,
                });
            };

            // TODO: support multiple DID bindings
            context.holder_did = did.into();
        }
    }

    Ok(())
}

async fn process(
    context: &Context, provider: impl Provider, request: &CredentialRequest,
) -> Result<CredentialResponse> {
    tracing::debug!("credential::process");

    // generate new nonce
    let mut state = context.state.clone();
    let Step::Token(mut token_state) = state.current_step else {
        return Err(Error::ServerError("Invalid token state".into()));
    };
    token_state.c_nonce = gen::nonce();
    token_state.c_nonce_expires_at = Utc::now() + Expire::Nonce.duration();
    state.current_step = Step::Token(token_state.clone());

    let mut response = CredentialResponse {
        c_nonce: Some(token_state.c_nonce.clone()),
        c_nonce_expires_in: Some(token_state.c_nonce_expires_in()),
        ..CredentialResponse::default()
    };
    let state_key: String;

    // if no VC is returned, issuance is deferred
    if let Some(credential) = create_vc(context, provider.clone(), request).await? {
        // sign credential (as jwt)
        let signer = SecOps::signer(&provider, &request.credential_issuer)
            .map_err(|e| Error::ServerError(format!("issue  resolving signer: {e}")))?;
        let jwt = vercre_w3c_vc::proof::create(Format::JwtVcJson, Payload::Vc(credential), signer)
            .await
            .map_err(|e| Error::ServerError(format!("issue creating proof: {e}")))?;

        state_key = token_state.access_token;
        response.credential = Some(Quota::One(Kind::String(jwt)));
    } else {
        let txn_id = gen::transaction_id();
        state.current_step = Step::Deferred(Deferred {
            transaction_id: txn_id.clone(),
            credential_request: request.clone(),
        });

        state_key = txn_id.clone();
        response.transaction_id = Some(txn_id);
    }

    // update state
    StateStore::put(&provider, &state_key, state.to_vec()?, state.expires_at)
        .await
        .map_err(|e| Error::ServerError(format!("issue saving state: {e}")))?;

    Ok(response)
}

// Attempt to generate a Verifiable Credential from information provided in the Credential
// Request. May return `None` if the credential is not ready to be issued because the request
// for Subject is pending.
async fn create_vc(
    context: &Context, provider: impl Provider, request: &CredentialRequest,
) -> Result<Option<VerifiableCredential>> {
    tracing::debug!("credential::create_vc");

    // get credential identifier and configuration
    let (identifier, config) = credential_configuration(context, request)?;
    let definition = credential_definition(request, &config);

    let Some(subject_id) = &context.state.subject_id else {
        return Err(Error::AccessDenied("holder not found".into()));
    };

    // get ALL claims for holder/credential
    let mut claims_resp = Subject::claims(&provider, subject_id, &identifier)
        .await
        .map_err(|e| Error::ServerError(format!("issue populating claims: {e}")))?;

    // defer issuance if claims are pending (approval?),
    if claims_resp.pending {
        return Ok(None);
    }

    // TODO: need to check authorized claims (claims in credential offer or authorization request)
    // retain ONLY requested (and mandatory) claims
    let cred_subj_def = &definition.credential_subject.unwrap_or_default();
    if let Some(req_cred_def) = &request.credential_definition {
        if let Some(req_cred_subj) = &req_cred_def.credential_subject {
            let mut claims = claims_resp.claims;
            claims.retain(|key, _| {
                req_cred_subj.get(key).is_some() || cred_subj_def.get(key).is_some()
            });
            claims_resp.claims = claims;
        }
    }

    let credential_issuer = &context.issuer_config.credential_issuer;

    // HACK: fix this (AW: why is this a hack?)
    let Some(types) = definition.type_ else {
        return Err(Error::ServerError("Credential type not set".into()));
    };
    let vc_id = format!("{credential_issuer}/credentials/{}", types[1].clone());

    let vc = VerifiableCredential::builder()
        .add_context(Kind::String(credential_issuer.clone() + "/credentials/v1"))
        // TODO: generate credential id
        .id(vc_id)
        .add_type(types[1].clone())
        .issuer(credential_issuer.clone())
        .add_subject(CredentialSubject {
            id: Some(context.holder_did.clone()),
            claims: claims_resp.claims,
        })
        .build()
        .map_err(|e| Error::ServerError(format!("issue building VC: {e}")))?;

    Ok(Some(vc))
}

fn credential_configuration(
    context: &Context, request: &CredentialRequest,
) -> Result<(String, CredentialConfiguration)> {
    match &request.credential_type {
        CredentialType::Identifier(identifier) => {
            let Some(config) =
                context.issuer_config.credential_configurations_supported.get(identifier)
            else {
                return Err(Error::InvalidCredentialRequest("credential is not supported".into()));
            };
            Ok((identifier.clone(), config.clone()))
        }
        CredentialType::Format(format) => {
            let Some(definition) = &request.credential_definition else {
                return Err(Error::InvalidCredentialRequest(
                    "credential definition not set".into(),
                ));
            };
            let Some(id_config) =
                context.issuer_config.credential_configurations_supported.iter().find(|(_, v)| {
                    &v.format == format && v.credential_definition.type_ == definition.type_
                })
            else {
                return Err(Error::InvalidCredentialRequest("credential is not supported".into()));
            };
            Ok((id_config.0.clone(), id_config.1.clone()))
        }
    }
}

/// Creates, stores, and returns new `c_nonce` and `c_nonce_expires`_in values
/// for use in `Error::InvalidProof` errors, as per specification.
async fn err_nonce(context: &Context, provider: &impl Provider) -> Result<(String, i64)> {
    // generate nonce and update state
    let mut state = context.state.clone();
    let Step::Token(mut token_state) = state.current_step else {
        return Err(Error::ServerError("token state not set".into()));
    };

    let c_nonce = gen::nonce();
    token_state.c_nonce.clone_from(&c_nonce);
    token_state.c_nonce_expires_at = Utc::now() + Expire::Nonce.duration();
    state.current_step = Step::Token(token_state.clone());

    StateStore::put(provider, &token_state.access_token, state.to_vec()?, state.expires_at)
        .await
        .map_err(|e| Error::ServerError(format!("issue saving state: {e}")))?;

    Ok((c_nonce, Expire::Nonce.duration().num_seconds()))
}

// Get the request's credential definition. If it does not exist, create it.
fn credential_definition(
    request: &CredentialRequest, config: &CredentialConfiguration,
) -> CredentialDefinition {
    tracing::debug!("credential::credential_definition");

    let mut definition =
        request.credential_definition.clone().unwrap_or_else(|| CredentialDefinition {
            context: None,
            type_: config.credential_definition.type_.clone(),
            credential_subject: config.credential_definition.credential_subject.clone(),
        });

    // add credential subject when not present
    if definition.credential_subject.is_none() {
        definition.credential_subject.clone_from(&config.credential_definition.credential_subject);
    };

    definition
}

#[cfg(test)]
mod tests {
    use assert_let_bind::assert_let;
    use insta::assert_yaml_snapshot as assert_snapshot;
    use serde_json::json;
    use vercre_test_utils::holder;
    use vercre_test_utils::issuer::{Provider, CLIENT_ID, CREDENTIAL_ISSUER, NORMAL_USER};
    use vercre_w3c_vc::proof::Verify;

    use super::*;
    use crate::state::Token;

    #[tokio::test]
    async fn credential_ok() {
        vercre_test_utils::init_tracer();

        let provider = Provider::new();
        let access_token = "ABCDEF";
        let c_nonce = "1234ABCD";
        let credentials = vec!["EmployeeID_JWT".into()];

        // set up state
        let mut state = State {
            expires_at: Utc::now() + Expire::Authorized.duration(),
            credential_identifiers: credentials,
            subject_id: Some(NORMAL_USER.into()),
            ..State::default()
        };

        state.current_step = Step::Token(Token {
            access_token: access_token.into(),
            c_nonce: c_nonce.into(),
            c_nonce_expires_at: Utc::now() + Expire::Nonce.duration(),
        });

        let ser = state.to_vec().expect("should serialize");
        StateStore::put(&provider, access_token, ser, state.expires_at).await.expect("state saved");

        // create CredentialRequest to 'send' to the app
        let claims = ProofClaims {
            iss: Some(CLIENT_ID.into()),
            aud: CREDENTIAL_ISSUER.into(),
            iat: Utc::now().timestamp(),
            nonce: Some(c_nonce.into()),
        };
        let jwt = jws::encode(Type::Proof, &claims, holder::Provider).await.expect("should encode");

        let body = json!({
            "format": "jwt_vc_json",
            "credential_definition": {
                "type": [
                    "VerifiableCredential",
                    "EmployeeIDCredential"
                ]
            },
            "proof":{
                "proof_type": "jwt",
                "jwt": jwt
            }
        });

        let mut request =
            serde_json::from_value::<CredentialRequest>(body).expect("request should deserialize");
        request.credential_issuer = CREDENTIAL_ISSUER.into();
        request.access_token = access_token.into();

        let response = credential(provider.clone(), &request).await.expect("response is valid");
        assert_snapshot!("response", &response, {
            ".credential" => "[credential]",
            ".c_nonce" => "[c_nonce]",
            ".c_nonce_expires_in" => "[c_nonce_expires_in]"
        });

        // verify credential
        let vc_quota = response.credential.expect("credential is present");
        let Quota::One(vc_kind) = vc_quota else {
            panic!("expected one credential")
        };

        let Payload::Vc(vc) = vercre_w3c_vc::proof::verify(Verify::Vc(&vc_kind), &provider)
            .await
            .expect("should decode")
        else {
            panic!("should be VC");
        };

        assert_snapshot!("vc", vc, {
            ".issuanceDate" => "[issuanceDate]",
            ".credentialSubject" => insta::sorted_redaction()
        });

        // token state should remain unchanged
        assert_let!(Ok(buf), StateStore::get(&provider, access_token).await);
        let state = State::try_from(buf).expect("state is valid");
        assert_snapshot!("state", state, {
            ".expires_at" => "[expires_at]",
            ".current_step.c_nonce"=>"[c_nonce]",
            ".current_step.c_nonce_expires_at" => "[c_nonce_expires_at]"
        });
    }

    #[tokio::test]
    async fn authorization_details() {
        vercre_test_utils::init_tracer();

        let provider = Provider::new();
        let access_token = "ABCDEF";
        let c_nonce = "1234ABCD";
        let credentials = vec!["EmployeeID_JWT".into()];

        // set up state
        let mut state = State {
            expires_at: Utc::now() + Expire::Authorized.duration(),
            credential_identifiers: credentials,
            subject_id: Some(NORMAL_USER.into()),
            ..State::default()
        };

        state.current_step = Step::Token(Token {
            access_token: access_token.into(),
            c_nonce: c_nonce.into(),
            c_nonce_expires_at: Utc::now() + Expire::Nonce.duration(),
        });

        let ser = state.to_vec().expect("should serialize");
        StateStore::put(&provider, access_token, ser, state.expires_at)
            .await
            .expect("state exists");

        let claims = ProofClaims {
            iss: Some(CLIENT_ID.into()),
            aud: CREDENTIAL_ISSUER.into(),
            iat: Utc::now().timestamp(),
            nonce: Some(c_nonce.into()),
        };
        let jwt = jws::encode(Type::Proof, &claims, holder::Provider).await.expect("should encode");

        let body = json!({
            "format": "jwt_vc_json",
            "credential_definition": {
                "type": ["VerifiableCredential", "EmployeeIDCredential"],
                "credentialSubject": {
                    "givenName": {},
                    "familyName": {},
                }
            },
            "proof":{
                "proof_type": "jwt",
                "jwt": jwt
            }
        });

        let mut request =
            serde_json::from_value::<CredentialRequest>(body).expect("request should deserialize");
        request.credential_issuer = CREDENTIAL_ISSUER.into();
        request.access_token = access_token.into();

        let response = credential(provider.clone(), &request).await.expect("response is valid");
        assert_snapshot!("ad-response", &response, {
            ".credential" => "[credential]",
            ".c_nonce" => "[c_nonce]",
            ".c_nonce_expires_in" => "[c_nonce_expires_in]"
        });

        // verify credential
        let vc_quota = response.credential.expect("credential is present");
        let Quota::One(vc_kind) = vc_quota else {
            panic!("expected one credential")
        };

        let Payload::Vc(vc) = vercre_w3c_vc::proof::verify(Verify::Vc(&vc_kind), &provider)
            .await
            .expect("should decode")
        else {
            panic!("should be VC");
        };

        assert_snapshot!("ad-vc", vc, {
            ".issuanceDate" => "[issuanceDate]",
            ".credentialSubject" => insta::sorted_redaction()
        });

        // token state should remain unchanged
        assert_let!(Ok(buf), StateStore::get(&provider, access_token).await);
        let state = State::try_from(buf).expect("state is valid");
        assert_snapshot!("ad-state", state, {
            ".expires_at" => "[expires_at]",
            ".current_step.c_nonce"=>"[c_nonce]",
            ".current_step.c_nonce_expires_at" => "[c_nonce_expires_at]"
        });
    }

    // #[tokio::test]
    // async fn credential_identifiers() {
    //     test_utils::init_tracer();

    //     let provider = Provider::new();
    //     let access_token = "ABCDEF";
    //     let c_nonce = "1234ABCD";
    //     let identifiers = vec!["EmployeeID_JWT".into()];

    //     // set up state
    //     let mut state = State::builder()
    //         .credential_issuer(CREDENTIAL_ISSUER.into())
    //         .expires_at(Utc::now() + Expire::Authorized.duration())
    //         .subject_id(Some(NORMAL_USER.into()))
    //         .build()
    //         .expect("should build state");

    //     state.token = Some(Token {
    //         access_token: access_token.into(),
    //         token_type: "Bearer".into(),
    //         c_nonce: c_nonce.into(),
    //         c_nonce_expires_at: Utc::now() + Expire::Nonce.duration(),
    //         ..Default::default()
    //     });

    //     StateStore::put(&provider, access_token, state.to_vec(), state.expires_at)
    //         .await
    //         .expect("state exists");

    //     // create CredentialRequest to 'send' to the app

    //     let claims = ProofClaims {
    //         iss: holder_provider::CLIENT_ID,
    //         aud: CREDENTIAL_ISSUER.into(),
    //         iat: Utc::now().timestamp(),
    //         nonce: c_nonce.into(),
    //     };
    //  let jwt = jose::encode(jose::Payload::Proof, &claims, holder_provider::Provider::new())
    //         .await
    //         .expect("should encode");

    //     let body = json!({
    //         "credential_requests":[{
    //             "credential_identifier": "EmployeeID_JWT",
    //             "proof":{
    //                 "proof_type": "jwt",
    //                 "jwt": signed_jwt
    //             }
    //         }]
    //     });

    //     let mut request = serde_json::from_value::<CredentialRequest>(body)
    //         .expect("request should deserialize");
    //     request.credential_issuer = CREDENTIAL_ISSUER.into();
    //     request.access_token = access_token.into();

    //     let response =
    //         Endpoint::new(provider.clone()).batch(&request).await.expect("response is valid");
    //     assert_snapshot!("ci-response", response, {
    //         ".credential_responses[0]" => "[credential]",
    //         ".c_nonce" => "[c_nonce]",
    //         ".c_nonce_expires_in" => "[c_nonce_expires_in]"
    //     });

    //     // verify credential
    //     assert!(response.credential_responses.len() == 1);
    //     let credential = response.credential_responses[0].credential.clone();

    //     let vc_val = credential.expect("VC is present");
    //     let token = serde_json::from_value::<String>(vc_val).expect("base64 encoded string");
    //     let Payload::Vc(vc) = proof::verify(&token, Verify::Vc).await.expect("should decode")
    //     else {
    //         panic!("should be VC");
    //     };

    //     assert_snapshot!("ci-vc_jwt", vc_jwt, {
    //         ".claims.iat" => "[iat]",
    //         ".claims.nbf" => "[nbf]",
    //         ".claims.vc.issuanceDate" => "[issuanceDate]",
    //         ".claims.vc.credentialSubject" => insta::sorted_redaction()
    //     });

    //     // token state should remain unchanged
    //     assert_let!(Ok(buf), StateStore::get(&provider, access_token).await);
    //     let state = State::try_from(buf).expect("state is valid");
    //     assert_snapshot!("ci-state", state, {
    //         ".expires_at" => "[expires_at]",
    //         ".token.c_nonce"=>"[c_nonce]",
    //         ".token.c_nonce_expires_at" => "[c_nonce_expires_at]"
    //     });
    // }
}
