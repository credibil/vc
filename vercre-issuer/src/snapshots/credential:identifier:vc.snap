---
source: vercre-issuer/src/credential.rs
assertion_line: 440
expression: vc
---
"@context":
  - "https://www.w3.org/2018/credentials/v1"
  - "http://vercre.io/credentials/v1"
id: "http://vercre.io/credentials/EmployeeIDCredential"
type:
  - VerifiableCredential
  - EmployeeIDCredential
issuer: "http://vercre.io"
issuanceDate: "[issuanceDate]"
credentialSubject:
  email: normal.user@example.com
  family_name: Person
  given_name: Normal
  id: "did:key:z6Mkj8Jr1rg3YjVWWhg7ahEYJibqhjBgZt1pDCbT4Lv7D4HX"
  proficiency: "3"
